addon_imps/storage/owncloud.py

```diff
@@ -0,0 +1,227 @@
+
+    def _parse_displayname(self, response_xml: str) -> str:
+        ns = {"d": "DAV:"}
+        root = ET.fromstring(response_xml)
+        displayname = root.find(".//d:displayname", ns)
+        if displayname is not None and displayname.text:
+            return displayname.text
+        else:
+            raise ValueError("displayname not found in response")
+
+    def _parse_item(self, response_xml: str, path: str) -> storage.ItemResult:
+        ns = {"d": "DAV:", "oc": "http://owncloud.org/ns"}
+        root = ET.fromstring(response_xml)
+        response_element = root.find("d:response", ns)
+        if response_element is None:
+            raise ValueError("No response element found in PROPFIND response")
+
+        resourcetype = response_element.find(".//d:resourcetype", ns)
+        if (
+            resourcetype is not None
+            and resourcetype.find("d:collection", ns) is not None
+        ):
+            item_type = storage.ItemType.FOLDER
+        else:
+            item_type = storage.ItemType.FILE
+
+        displayname_element = response_element.find(".//d:displayname", ns)
+        if displayname_element is not None and displayname_element.text:
+            displayname = displayname_element.text
+        else:
+            displayname = path.rstrip("/").split("/")[-1]
+
+        item_result = storage.ItemResult(
+            item_id=_make_item_id(item_type, path),
+            item_name=displayname,
+            item_type=item_type,
+        )
+        return item_result
+
+    def _parse_items(
+        self,
+        response_xml: str,
+        base_path: str,
+        item_type: storage.ItemType | None = None,
+    ) -> typing.Iterator[storage.ItemResult]:
+        ns = {"d": "DAV:", "oc": "http://owncloud.org/ns"}
+        root = ET.fromstring(response_xml)
+        for response_element in root.findall("d:response", ns):
+            href_element = response_element.find("d:href", ns)
+            if href_element is None or not href_element.text:
+                continue
+            href = href_element.text
+            path = self._href_to_path(href)
+
+            if path.rstrip("/") == base_path.rstrip("/"):
+                continue
+
+            resourcetype = response_element.find(".//d:resourcetype", ns)
+            if (
+                resourcetype is not None
+                and resourcetype.find("d:collection", ns) is not None
+            ):
+                _item_type = storage.ItemType.FOLDER
+            else:
+                _item_type = storage.ItemType.FILE
+
+            if item_type is not None and _item_type != item_type:
+                continue
+
+            displayname_element = response_element.find(".//d:displayname", ns)
+            if displayname_element is not None and displayname_element.text:
+                displayname = displayname_element.text
+            else:
+                displayname = path.rstrip("/").split("/")[-1]
+
```

Comment:
it seems that these lines are nearly identical to 134-146, if it isn't much trouble, please make it more DRY